File: crustdesk-macos-classic/host/test_crypto.py
#!/usr/bin/env python3
"""Check the vendored libsodium subset against PyNaCl, byte for byte."""
import subprocess
import sys
from nacl.bindings import (crypto_box, crypto_scalarmult_base, crypto_secretbox,
                           crypto_sign, crypto_sign_seed_keypair)

exe = sys.argv[1]
seed = bytes((i * 7 + 1) & 255 for i in range(32))
bsk = bytes((i * 13 + 5) & 255 for i in range(32))
key = bytes((i * 3 + 9) & 255 for i in range(32))
nonce = bytes(range(24))
msg = b"C-Desk-Vint signs this"

pk, sk = crypto_sign_seed_keypair(seed)
bpk = crypto_scalarmult_base(bsk)
# A peer (sodiumoxide on the client) boxes a secretbox key to our box key.
peer_sk = bytes((i * 29 + 3) & 255 for i in range(32))
peer_pk = crypto_scalarmult_base(peer_sk)
sym = bytes(range(100, 132))
boxed = crypto_box(sym, bytes(24), bpk, peer_sk)

out = subprocess.run([exe, peer_pk.hex(), boxed.hex()], capture_output=True, text=True).stdout
got = dict(line.split(' ', 1) for line in out.strip().splitlines())
want = {
    'sign_pk': pk.hex(),
    'signed': crypto_sign(msg, sk).hex(),
    'box_pk': bpk.hex(),
    'secretbox': crypto_secretbox(msg, nonce, key).hex(),
    'opened': sym.hex(),
}
fails = [k for k in want if got.get(k) != want[k]]
for k in fails:
    print('FAIL', k, '\n  got ', got.get(k), '\n  want', want[k])
print('FAIL' if fails else 'PASS: crypto matches libsodium (sign, box, secretbox)')
sys.exit(1 if fails else 0)

File: crustdesk-macos-classic/host/test_crypto.c
/* The vendored libsodium subset, driven with fixed inputs; host/test_crypto.py
 * checks every output against PyNaCl (the system libsodium). */
#include "sodium/crypto_box.h"
#include "sodium/crypto_secretbox.h"
#include "sodium/crypto_sign_ed25519.h"

#include <stdio.h>
#include <string.h>

static void hex(const char *label, const unsigned char *b, unsigned long long n)
{
    unsigned long long i;
    printf("%s ", label);
    for (i = 0; i < n; i++)
        printf("%02x", b[i]);
    printf("\n");
}

int main(int argc, char **argv)
{
    unsigned char seed[32], pk[32], sk[64], sm[64 + 64], bpk[32], bsk[32], key[32], nonce[24];
    unsigned char c[256], m[256];
    unsigned long long n;
    const char *msg = "C-Desk-Vint signs this";
    int i;
    for (i = 0; i < 32; i++) {
        seed[i] = (unsigned char)(i * 7 + 1);
        bsk[i] = (unsigned char)(i * 13 + 5);
        key[i] = (unsigned char)(i * 3 + 9);
    }
    for (i = 0; i < 24; i++)
        nonce[i] = (unsigned char)i;

    crypto_sign_ed25519_seed_keypair(pk, sk, seed);
    hex("sign_pk", pk, 32);
    crypto_sign_ed25519(sm, &n, (const unsigned char *)msg, strlen(msg), sk);
    hex("signed", sm, n);

    crypto_scalarmult_curve25519_base(bpk, bsk);
    hex("box_pk", bpk, 32);

    crypto_secretbox_easy(c, (const unsigned char *)msg, strlen(msg), nonce, key);
    hex("secretbox", c, strlen(msg) + 16);

    /* argv[1]: a box made by PyNaCl for us: open it. */
    if (argc > 2) {
        unsigned char peer[32], boxed[128];
        size_t bl = strlen(argv[2]) / 2;
        for (i = 0; i < 32; i++)
            sscanf(argv[1] + 2 * i, "%2hhx", &peer[i]);
        for (i = 0; i < (int)bl; i++)
            sscanf(argv[2] + 2 * i, "%2hhx", &boxed[i]);
        memset(nonce, 0, sizeof nonce);
        if (crypto_box_open_easy(m, boxed, bl, nonce, peer, bsk) == 0)
            hex("opened", m, bl - 16);
        else
            printf("opened FAILED\n");
    }
    return 0;
}
